### DISCUSSION TITLE
how to grant chainStorage access to a contract?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#6839] [Q&A] [answered]
As noted in https://github.com/Agoric/agoric-sdk/issues/5687#issuecomment-1351934866 , we discussed this in Dec 14 office hours. Someone who watched that recording asked how to write their own version of `startPSM.js` in order to give the their contract access to a storageNode.

[chainStorage-proposal.js](https://gist.github.com/dckc/1230bf4a62530b049c53ec7740844033#file-chainstorage-proposal-js) (along with [bakeSale-permit.json](https://gist.github.com/dckc/1230bf4a62530b049c53ec7740844033#file-bakesale-permit-json) ) is a quick sketch. It's un-tested.

cc @Tyrosine22 @turadg

**Comment by carlos-kryha** [comment, marked as answer]
Looks like it works on local testnet, I've put together a document with the steps [here](https://nickel-infinity-3bd.notion.site/Deploying-to-Agoric-local-testnet-with-storageNode-powers-a88ed1a89b4f4cd08d5f7bfab60ee9cd) for anyone interested. 

Thanks for the example!

**Comment by dckc** [comment]
## agoricNames beyond reserved names

As we discussed, testing showed that not only does this idiom result in a static error, it doesn't actually work at runtime for names that aren't in a hard-coded "reserved" list:

```js
    instance: {
      // @ts-expect-error bakeSaleKit isn't declared in vats/src/core/types.js
      produce: { [contractInfo.instanceName]: produceInstance },
```

To register arbitrary names...

```js
  // Add agoricNamesAdmin to the consume
  const {
    consume: { agoricNamesAdmin, board, chainStorage, zoe },
    // @ts-expect-error bakeSaleKit isn't declared in vats/src/core/types.js
    produce: { bakeSaleKit },
    instance: {
      // @ts-expect-error bakeSaleKit isn't declared in vats/src/core/types.js
      produce: { [contractInfo.instanceName]: produceInstance },
    },
  } = powers;
```

And the "Share instance widely" code becomes...

```
  // Share instance widely via E(agoricNames).lookup('instance', 'bakeSaleAgent')
  const instanceAdmin = E(agoricNamesAdmin).lookupAdmin('instance');
  await E(instanceAdmin).update('bakeSaleAgent', facets.instance);
```

You asked about constraints/conventions in this area... whether sharing the publicFacet directly would work...

the [AgoricContractInvitationSpec](https://github.com/Agoric/agoric-sdk/blob/f447e3a93d1cd19d73b1310158bc4eea4f8bd824/packages/smart-wallet/src/invitations.js#L21-L29) type documentation doesn't say so, but the smartWallet [makeInvitationsHelper](https://github.com/Agoric/agoric-sdk/blob/f447e3a93d1cd19d73b1310158bc4eea4f8bd824/packages/smart-wallet/src/invitations.js#L67) calls `E(zoe).getPublicFacet(instance)` between the `instancePath` and the `callPipe`. So if the `instancePath` led directly to a public facet, the call to `E(zoe).getPublicFacet)` would fail.

So for use with the smart wallet, it has to be an instance that goes into agoricNames.

**Comment by dckc** [comment]
> it doesn't actually work at runtime for names that aren't in a hard-coded "reserved" list

I'm fixing this, btw...

https://github.com/Agoric/agoric-sdk/blob/b399893ca65f2942feeffee715ea03a81020a8dc/packages/vats/test/test-name-hub-published.js#L71

ref

 - #4548 
 - #7579 

p.s. landed in mainnet1B:

https://github.com/Agoric/agoric-sdk/blob/ba3b776444425bbcb275ce918794d9259f8e15da/packages/vats/test/test-name-hub-published.js#L51


 - #7604

**Comment by 0xpatrickdev** [comment]
What does extending the bakeSale example to `chainTimerService` look like? 

I added `"chainTimerService": "timer"` to `permit.json` and set `noTerms` to `harden({ timer: chainTimerService })`, but SwingSet doesn't seem to like this (`Error#1: value for instanceRecord is not durable`).

Looking at the Inter econ-behaviors as a reference, it seems we need to resolve chainTimerService with `deeplyFulfilledObject`, an imported module. Is the pattern here to deploy a second intermediary bundle, startBakeSale.js, to handle this setup and reference it in `bakeSale-proposal.js`?

**Comment by dckc** [comment]
The `chainTimerService` you're getting is a promise (all properties of `powers` are promises). Try resolving it (using `await` or the like) before using it in the terms of `startInstance`.

**Reply by 0xpatrickdev** [reply]
:homer_doh: that works - thank you!
